pkg/internal/itest/testpeer/generator.go

```diff
@@ -259,3 +329,89 @@ func StartAndWaitForReady(ctx context.Context, manager datatransfer.Manager) err
 		return err
 	}
 }
+
+func HttpHandler(ctx context.Context, lsys linking.LinkSystem) func(http.ResponseWriter, *http.Request) {
+	return func(res http.ResponseWriter, req *http.Request) {
+		urlPath := strings.Split(req.URL.Path, "/")[1:]
+
+		// validate CID path parameter
+		cidStr := urlPath[1]
+		rootCid, err := cid.Parse(cidStr)
+		if err != nil {
+			http.Error(res, fmt.Sprintf("Failed to parse CID path parameter: %s", cidStr), http.StatusBadRequest)
+			return
+		}
+
+		// Grab unixfs path if it exists
+		unixfsPath := ""
+		if len(urlPath) > 2 {
+			unixfsPath = "/" + strings.Join(urlPath[2:], "/")
+		}
+
+		// Parse car scope and use it to get selector
+		carScope := types.CarScopeAll
+		if req.URL.Query().Has("car-scope") {
+			switch req.URL.Query().Get("car-scope") {
+			case "all":
+			case "file":
+				carScope = types.CarScopeFile
+			case "block":
+				carScope = types.CarScopeBlock
+			default:
+				http.Error(res, fmt.Sprintf("Invalid  car-scope parameter: %s", req.URL.Query().Get("car-scope")), http.StatusBadRequest)
+				return
+			}
+		}
+		selNode := unixfsnode.UnixFSPathSelectorBuilder(unixfsPath, carScope.TerminalSelectorSpec(), false)
+		sel, err := selector.CompileSelector(selNode)
+		if err != nil {
+			http.Error(res, fmt.Sprintf("Failed to compile selector from car-scope: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		// Write to response writer
+		carWriter, err := storage.NewWritable(res, []cid.Cid{rootCid}, car.WriteAsCarV1(true), car.AllowDuplicatePuts(false))
+		if err != nil {
+			http.Error(res, fmt.Sprintf("Failed to create car writer: %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		// Extend the StorageReadOpener func to write to the carWriter
+		originalSRO := lsys.StorageReadOpener
+		lsys.StorageReadOpener = func(lc linking.LinkContext, lnk datamodel.Link) (io.Reader, error) {
+			r, err := originalSRO(lc, lnk)
+			if err != nil {
+				return nil, err
+			}
+			byts, err := io.ReadAll(r)
+			if err != nil {
+				return nil, err
+			}
+			err = carWriter.Put(ctx, lnk.(cidlink.Link).Cid.KeyString(), byts)
+			if err != nil {
+				return nil, err
+			}
+
+			return bytes.NewReader(byts), nil
+		}
+
+		rootNode, err := lsys.Load(linking.LinkContext{}, cidlink.Link{Cid: rootCid}, basicnode.Prototype.Any)
```

Comment:
So, this is kind of subtle, and I tripped over this twice yesterday cause it's easy to forget—the prototype argument (last) matters a lot here because we care about special-casing unixfs. What you're going to discover when you expand the test cases is that they fail for paths and for `car-scope=file` because it'll hit this: https://github.com/ipfs/go-unixfsnode/blob/main/reification.go#L27-L30

My fault cause the code you copied and pasted didn't have this properly worked out cause the tests that were using it are not exercising unixfs.

A dag-pb block _must_ use the dag-pb prototype so it becomes a `dagpb.PBNode` type, then go-unixfsnode finds it acceptable that it _might_ be unixfs (there's a spec argument about this going on here: https://github.com/ipfs/go-unixfsnode/pull/27). So, you have to make sure that any loading of bytes into `ipld.Node` form goes through the dag-pb prototype _if_ the CID says that it's dag-pb. There's a couple of approaches for this particular bit of code (both of which I used in my current https://github.com/filecoin-project/lassie/pull/222, a different one for tests).

First way is explicit and kind of manual; inspect the CID and branch on which prototype you feed in here:

```go
		var proto datamodel.NodePrototype = basicnode.Prototype.Any
		if rootCid.Prefix().Codec == cid.DagProtobuf {
			proto = dagpb.Type.PBNode
		}
		rootNode, err := lsys.Load(linking.LinkContext{}, cidlink.Link{Cid: rootCid}, proto)
```

The other way is to use a "chooser":

```go
		protoChooser = dagpb.AddSupportToChooser(basicnode.Chooser)
		proto, err := protoChooser(lnk, lc)
		if err != nil {
				// handle
		}
		rootNode, err := lsys.Load(linking.LinkContext{}, cidlink.Link{Cid: rootCid}, proto)
```

Essentially it's something like this: `basicnode.Chooser` is always just going to give you a `basicnode.Prototype.Any`, but you can wrap it with `dagpb.AddSupportToChooser` such that it'll inspect the CID and give you the dag-pb prototype instead.

The reason you should probably opt for the second option is that you're going to need `protoChooser` anyway down below in the traversal configuration for `LinkTargetNodePrototypeChooser`.

Reply:
+1 here

Reply:
IPLD is weird and complicated